File: test/language/sql_tables/cpp98/SubtypedBoolFieldTableTest.cpp
#include <cstdio>
#include <string>
#include <fstream>

#include "gtest/gtest.h"

#include "sql_tables/TestDb.h"

namespace sql_tables
{
namespace subtyped_bool_field_table
{

class SubtypedBoolFieldTableTest : public ::testing::Test
{
public:
    SubtypedBoolFieldTableTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new sql_tables::TestDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~SubtypedBoolFieldTableTest()
    {
        delete m_database;
    }

protected:
    static void fillRow(SubtypedBoolFieldTableRow& row, size_t i)
    {
        row.setId(static_cast<int32_t>(i));
        row.setBoolField(i % 2 == 0);
    }

    static void fillRows(std::vector<SubtypedBoolFieldTableRow>& rows)
    {
        rows.clear();
        for (size_t i = 0; i < NUM_ROWS; ++i)
        {
            SubtypedBoolFieldTableRow row;
            fillRow(row, i);
            rows.push_back(row);
        }
    }

    static void checkRow(const SubtypedBoolFieldTableRow& row1, const SubtypedBoolFieldTableRow& row2)
    {
        ASSERT_EQ(row1.getId(), row2.getId());
        ASSERT_EQ(row1.getBoolField(), row2.getBoolField());
    }

    static void checkRows(const std::vector<SubtypedBoolFieldTableRow>& rows1,
            const std::vector<SubtypedBoolFieldTableRow>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
            checkRow(rows1[i], rows2[i]);
    }

    static const char DB_FILE_NAME[];
    static const size_t NUM_ROWS;

    sql_tables::TestDb* m_database;
};

const char SubtypedBoolFieldTableTest::DB_FILE_NAME[] = "subtyped_bool_field_table_test.sqlite";
const size_t SubtypedBoolFieldTableTest::NUM_ROWS = 5;

TEST_F(SubtypedBoolFieldTableTest, readWithoutCondition)
{
    SubtypedBoolFieldTable& table = m_database->getSubtypedBoolFieldTable();

    std::vector<SubtypedBoolFieldTableRow> rows;
    fillRows(rows);
    table.write(rows);

    std::vector<SubtypedBoolFieldTableRow> readRows;
    table.read(readRows);
    checkRows(rows, readRows);
}

} // namespace subtyped_bool_field_table
} // namespace sql_tables

File: test/language/sql_tables/java/sql_tables/subtyped_bool_field_table/SubtypedBoolFieldTableTest.java
package sql_tables.subtyped_bool_field_table;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import sql_tables.TestDb;

import zserio.runtime.ZserioError;

public class SubtypedBoolFieldTableTest
{
    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @BeforeEach
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new TestDb(file.toString());
        database.createSchema();
    }

    @AfterEach
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final SubtypedBoolFieldTable table = database.getSubtypedBoolFieldTable();

        final List<SubtypedBoolFieldTableRow> writtenRows = new ArrayList<SubtypedBoolFieldTableRow>();
        fillRows(writtenRows);
        table.write(writtenRows);

        final List<SubtypedBoolFieldTableRow> readRows = table.read();
        checkRows(writtenRows, readRows);
    }

    private static void fillRows(List<SubtypedBoolFieldTableRow> rows)
    {
        for (int i = 0; i < NUM_ROWS; ++i)
        {
            rows.add(createRow(i));
        }
    }

    private static SubtypedBoolFieldTableRow createRow(int i)
    {
        final SubtypedBoolFieldTableRow row = new SubtypedBoolFieldTableRow();
        row.setId(i);
        row.setBoolField(i % 2 == 0);

        return row;
    }

    private static void checkRows(List<SubtypedBoolFieldTableRow> rows1, List<SubtypedBoolFieldTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkRow(rows1.get(i), rows2.get(i));
    }

    private static void checkRow(SubtypedBoolFieldTableRow row1, SubtypedBoolFieldTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getBoolField(), row2.getBoolField());
    }

    private static final int    NUM_ROWS = 5;
    private static final String FILE_NAME = "subtyped_bool_field_table_test.sqlite";

    private final File file = new File(FILE_NAME);
    private TestDb database = null;
}
